### PR TITLE
Exploration of London smart meter data

### LINKED ISSUE #2: Exploration of London smart meter data

### PR DESCRIPTION
**Pull request by ch-williamson**
---

# Description

This PR features all of the scripts for performing exploratory analysis of the London smart meter data.

Fixes #2

# Instructions for Reviewer

The core functionality is producing plots of the energy profile clusters - this can be done by running `analysis/clustering.py` as a script, so please check that this works for you. As well as this it would be great if you could check my logic in the data aggregations and processing.

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

**Comment by ch-williamson** [comment]
Hello @sofiapinto! This is ready for review now - all the info is in the docs, but let me know if anything doesn't work or if you have any questions :)

File: README.md [hunk 1]
```diff
@@ -1,12 +1,75 @@
...
 - Meet the data science cookiecutter [requirements](http://nestauk.github.io/ds-cookiecutter/quickstart), in brief:
   - Install: `direnv` and `conda`
 - Run `make install` to configure the development environment:
   - Setup the conda environment
   - Configure `pre-commit`
+- Download the data:
+  - `make inputs-pull` will pull the zipped data from S3 and put it in `/inputs` - you can either unzip it manually or just run the scripts in `getters` which will do it automatically
+  - Alternatively, download the data from [Kaggle](https://www.kaggle.com/datasets/jeanmidev/smart-meters-in-london)
+- Perform additional setup in order to save plots:
+  - Follow the instructions [here](https://github.com/altair-viz/altair_saver/#nodejs)
```
Comment (sofiapinto) [inline README.md]: I would add a few lines, to help the user:

```suggestion
- Meet the data science cookiecutter [requirements](http://nestauk.github.io/ds-cookiecutter/quickstart), in brief:
  - Install: `direnv` and `conda`
- Clone the repo: `git clone git@github.com:nestauk/asf_smart_meter_exploration.git`
- Navigate to the repo folder
- Checkout the correct branch if not working on `dev`
- Run `make install` to configure the development environment:
  - Setup the conda environment
  - Configure `pre-commit`
- Run `direnv allow`;
- Activate conda environment: `conda activate asf_smart_meter_exploration`
- Download the data:
  - `make inputs-pull` will pull the zipped data from S3 and put it in `/inputs` - you can either unzip it manually or just run the scripts in `getters` which will do it automatically
  - Alternatively, download the data from [Kaggle](https://www.kaggle.com/datasets/jeanmidev/smart-meters-in-london)
- Perform additional setup in order to save plots:
  - Follow the instructions [here](https://github.com/altair-viz/altair_saver/#nodejs)
```

File: README.md [hunk 2]
```diff
@@ -1,12 +1,75 @@
...
+
 ## Setup
 
 - Meet the data science cookiecutter [requirements](http://nestauk.github.io/ds-cookiecutter/quickstart), in brief:
   - Install: `direnv` and `conda`
 - Run `make install` to configure the development environment:
   - Setup the conda environment
   - Configure `pre-commit`
+- Download the data:
+  - `make inputs-pull` will pull the zipped data from S3 and put it in `/inputs` - you can either unzip it manually or just run the scripts in `getters` which will do it automatically
```
Comment (sofiapinto) [inline README.md]: I would rephrase this since unzipping manually is different from running `asf_smart_meter_exploration/getters/process_raw_data.py` - you do some processing to the data in `process_raw_data.py`

File: asf_smart_meter_exploration/getters/process_raw_data.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import zipfile
+
+from asf_smart_meter_exploration import config, PROJECT_DIR
+
+meter_data_zip_path = PROJECT_DIR / config["meter_data_zip_path"]
+meter_data_folder_path = PROJECT_DIR / config["meter_data_folder_path"]
+meter_data_merged_file_path = PROJECT_DIR / config["meter_data_merged_file_path"]
+
+
+def produce_all_properties_df():
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/getters/process_raw_data.py]: I would probably split this into two functions:
- one to unzip data
- another to do the actual processing

And would also maybe add this to the pipeline folder instead of the getters.

I also left some suggestions below that might make the function a little faster, as it is a bit slow at the moment. So might be worth giving it a try to see if it improves time :)

File: asf_smart_meter_exploration/getters/process_raw_data.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            raise FileNotFoundError(
+                "Neither raw data folder nor zip file were found. Please check file location or redownload data from S3."
+            )
+        else:
+            print("Unzipping...")
+            with zipfile.ZipFile(meter_data_zip_path, "r") as zip_ref:
+                zip_ref.extractall("inputs")
+            print("Unzipped!")
+
+    halfhourly_dataset = []
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/getters/process_raw_data.py]: You can actually define this as a pandas DataFrame and work with only dataframes.
```suggestion
    halfhourly_dataset = pd.DataFrame()
```

File: asf_smart_meter_exploration/getters/process_raw_data.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            index_col="tstp",
+            parse_dates=True,
+            low_memory=False,
+        )
+        df_temp["file_name"] = file_name.split(".")[0]
+        df_temp = df_temp.replace("Null", np.nan).dropna()
+        df_temp["energy(kWh/hh)"] = df_temp["energy(kWh/hh)"].astype("float")
+        halfhourly_dataset.append(df_temp)
+
+    halfhourly_dataset = pd.concat(halfhourly_dataset, axis=0)
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/getters/process_raw_data.py]: Might make it faster to work with pandas dataframes instead of lists:
```suggestion
        halfhourly_dataset = pd.concat([halfhourly_dataset, df_temp])
```

File: asf_smart_meter_exploration/getters/process_raw_data.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    for file_name in tqdm.tqdm(os.listdir(meter_data_folder_path)):
+        df_temp = pd.read_csv(
+            os.path.join(meter_data_folder_path, file_name),
+            index_col="tstp",
+            parse_dates=True,
+            low_memory=False,
+        )
+        df_temp["file_name"] = file_name.split(".")[0]
+        df_temp = df_temp.replace("Null", np.nan).dropna()
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/getters/process_raw_data.py]: So we remove any instance with at least 1 missing value from the analysis? Do we know if this corresponds to a big proportion of the data?

File: asf_smart_meter_exploration/getters/process_raw_data.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            )
+        else:
+            print("Unzipping...")
+            with zipfile.ZipFile(meter_data_zip_path, "r") as zip_ref:
+                zip_ref.extractall("inputs")
+            print("Unzipped!")
+
+    halfhourly_dataset = []
+
+    for file_name in tqdm.tqdm(os.listdir(meter_data_folder_path)):
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/getters/process_raw_data.py]: Very cool to have the progress bar :D !
I would add a print before to let the user know we are processing the data. Also, I think if you define the list of folders before the for loop, it might make the process a little faster.
```suggestion
    print("Processing the data...")
    folder_names = os.listdir(meter_data_folder_path)
    for file_name in tqdm.tqdm(folder_names):
```

File: README.md [hunk 8]
```diff
@@ -1,12 +1,75 @@
...
+```mermaid
+  graph TD;
+      getters/process_daw_data.py;
+      getters/get_processed_data.py-->pipeline/data_aggregation.py;
+      pipeline/data_aggregation.py-->analysis/clustering.py;
+      utils/clustering_utils.py-->analysis/inertia_plots.py;
+      utils/clustering_utils.py-->analysis/clustering.py;
+      utils/plotting_utils.py-->analysis/clustering.py;
+      analysis/clustering.py-->analysis/inertia_plots.py;
+```
```
Comment (sofiapinto) [inline README.md]: very cool dependency map! Thought this was an image :D Also correcting a small typo below.

Another suggestion: why not connect `process_raw_data` to `get_processed_data`? By doing something like "If processed data in output folder read data from there, else run `process_raw_data`"

```suggestion
```mermaid
  graph TD;
      getters/process_raw_data.py;
      getters/get_processed_data.py-->pipeline/data_aggregation.py;
      pipeline/data_aggregation.py-->analysis/clustering.py;
      utils/clustering_utils.py-->analysis/inertia_plots.py;
      utils/clustering_utils.py-->analysis/clustering.py;
      utils/plotting_utils.py-->analysis/clustering.py;
      analysis/clustering.py-->analysis/inertia_plots.py;
```
```

File: asf_smart_meter_exploration/getters/process_raw_data.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        halfhourly_dataset.append(df_temp)
+
+    halfhourly_dataset = pd.concat(halfhourly_dataset, axis=0)
+
+    # Structure dataframe so that index is timestamps and columns are households
+    df_output = (
+        halfhourly_dataset.groupby(["tstp", "LCLid"])["energy(kWh/hh)"]
+        .mean(numeric_only=True)
+        .unstack()
+    )
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/getters/process_raw_data.py]: I checked on Kaggle what LCLid means. Maybe it would be good to write that somewhere. Simplest way would be to write that on your comment above.
```suggestion
    # Structure dataframe so that index is timestamps and columns are households (originally in the LCLid variable)
    df_output = (
        halfhourly_dataset.groupby(["tstp", "LCLid"])["energy(kWh/hh)"]
        .mean(numeric_only=True)
        .unstack()
    )
```

File: requirements.txt [hunk 10]
```diff
@@ -0,0 +1,9 @@
+pandas
+numpy
+tqdm
+altair
+scikit-learn
+matplotlib
+holidays
+altair_saver
+selenium==4.2.0
```
Comment (sofiapinto) [inline requirements.txt:9]: What is selenium used for?

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+
+meter_data = get_meter_data()
+
+
+def get_average_usage(meter_data=meter_data, normalised=False, cumulative=False):
+    """For each household, get average usage for each half-hour of the day.
+
+    Args:
+        meter_data (pd.DataFrame, optional): _description_. Defaults to meter_data.
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py]: The description for `meter_data` is missing

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+            the proportion used in each half-hour rather than the amount in kWh.
+            Defaults to False.
+        cumulative (bool, optional): Whether to return the cumulative sums of the values instead.
+            Defaults to False.
+
+    Returns:
+        pd.DataFrame: Average usage data.
+    """
+
+    hh_averages = meter_data.groupby("time").mean(numeric_only=True).dropna(axis=1).T
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py]: I was thinking about the missing values... I wonder why do we have missing values? Some disconnection problem? No one living in the house? Or should these be zeros instead?

I checked, and we only "lose" one household by doing dropna() here - which means that, for at least one time of time of day this household did not have any readings in the ~3 years the data was collected. 
However, it takes only one reading per time for this `mean()` function to work (as `mean()` ignores missing values) and for us to consider the household averages. This doesn't really matter for the purpose of this exploration exercise, but in the future we should be careful when doing this: perhaps if the % of missing values is very high (if for at least one time of day, the % of missing values for that household is above a certain x), then we do not consider the household for analysis.

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+    if normalised:
+        hh_averages = hh_averages.div(hh_averages.sum(axis=1), axis=0).dropna(axis=0)
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py:32]: I don't think we need to `dropna()` here, right? :)

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+    meter_data = meter_data.drop(
+        ["tstp", "day_of_week", "day_type", "bank_holiday"], axis=1, errors="ignore"
+    )
+
+    # Calculate means for each half-hour and day type pair
+    meter_data_daytypes = (
+        meter_data.groupby(["weekend_or_bank_holiday", "time"])
+        .mean(numeric_only=True)
+        .dropna(axis=1)
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py:76]: Why do we need the `dropna()` here?

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+        .T
+    )
+
+    if normalise:
+        meter_data_daytypes_norm = meter_data_daytypes.div(
+            meter_data_daytypes.groupby(level=0, axis=1).transform("sum"), axis=0
+        ).dropna(axis=0)
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py:83]: Do we really need the `dropna()`?

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+    meter_data["weekend_or_bank_holiday"] = (
+        meter_data["day_type"] == "weekend"
+    ) | meter_data["bank_holiday"]
+
+    meter_data = meter_data.drop(
+        ["tstp", "day_of_week", "day_type", "bank_holiday"], axis=1, errors="ignore"
+    )
+
+    # Calculate means for each half-hour and day type pair
+    meter_data_daytypes = (
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py]: Very elegant solution to combine the values for weekend/non-weekends :)

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+
+    season_dict = {
+        "winter": 0,
+        "spring": 1,
+        "summer": 2,
+        "autumn": 3,
+    }
+
+    meter_data["season"] = (
+        meter_data.tstp.dt.month // 3 % 4
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py]: like it, very elegant :)

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+    )  # quick way of getting season number
+    seasonal_aves = (
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py]: maybe naming it `seasonal_avgs` or `seasonal_averages` instead of `seasonal_aves`, to make it easier to understand? :)

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+    )
+
+    if season_2 != "spring and autumn":
+        seasonal_diff = (
+            seasonal_aves.loc[season_dict[season_1]]
+            - seasonal_aves.loc[season_dict[season_2]]
+        ).T
+    else:
+        spring_autumn_ave = (seasonal_aves.loc[1] + seasonal_aves.loc[3]) / 2
+        seasonal_diff = (seasonal_aves.loc[season_dict[season_1]] - spring_autumn_ave).T
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py]: Why not give the opportunity to compare these to the remaining two seasons? And maybe also allowing the remaining two seasons to be compared to any other?

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 20]
```diff
@@ -0,0 +1,177 @@
+# File: asf_smart_meter_exploration/pipeline/data_aggregation.py
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py:1]: Different things happen when I open a jupyter notebook and start calling these functions in different orders:

If I do 
```
from asf_smart_meter_exploration.pipeline.data_aggregation import get_average_usage_daytypes
data = get_average_usage_daytypes()
```
I get something different than when I do:
```
from asf_smart_meter_exploration.pipeline.data_aggregation import get_season_diff
aux = get_season_diff()

from asf_smart_meter_exploration.pipeline.data_aggregation import get_average_usage_daytypes
data = get_average_usage_daytypes()
```

The difference being, `data` has an extra line (for `season`) when I run the second set of lines of code. I think this is due to the fact that you have `meter_data = get_meter_data()` at the top of the script, outside of the functions. This is no problem as the important values in the dataframes don't seem to be changing. I think you just need to make sure that you only have values starting with "MAC" as households before returning the output dataframe of all the functions below. Otherwise, you might have the variables `day_of_week`, `bank_holiday`, `weekend_or_bank_holiday`, `season`, etc appearing as "households"  when you run these functions after one another.

File: asf_smart_meter_exploration/utils/plotting_utils.py [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+        chart (alt.Chart): Any Altair chart.
+
+    Returns:
+        alt.Chart: Chart with adjusted properties.
+    """
+    return (
+        chart.properties(width=800, height=300)
+        .configure_axis(labelFontSize=20, titleFontSize=20)
+        .configure_legend(titleFontSize=18, labelFontSize=15)
+        .configure_range(category={"scheme": "dark2"})
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/utils/plotting_utils.py]: Not sure if this is ever due to change or if it even makes sense to you, but you could have the font sizes, width, height, etc set up in base.yaml and imported here. This way, if you ever want to change these, you know that all properties due to change will be in base.yaml and easier to find.

File: asf_smart_meter_exploration/utils/plotting_utils.py [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+    ax.legend(title="Cluster", loc="upper right")
+    ax.set_xlabel("Time")
+    ax.set_ylabel(ylabel)
+    ax.set_xlim(datetime.time(0, 0, 0), datetime.time(23, 30, 00))
+    ax.set_ylim(ymin, ymax)
+    if normalised:
+        ax.yaxis.set_major_formatter(mtick.PercentFormatter(1, decimals=0))
+
+    ax.set_xticks(["00:00:00", "06:00:00", "12:00:00", "18:00:00"])
+    ax.locator_params(axis="y", nbins=4)
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/utils/plotting_utils.py]: Not sure why, but some plots have y axes with 3 bins instead of 4 (e.g. `winter_summer_diff_observations.png`)

File: asf_smart_meter_exploration/utils/plotting_utils.py [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+
+    plt.savefig(
+        cluster_plot_folder_path / (filename_infix + "_observations" + plot_suffix),
+        dpi=100,
+    )
+    plt.clf()
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/utils/plotting_utils.py:112]: I never use this function to clear the figure - do we really need it?

File: asf_smart_meter_exploration/utils/plotting_utils.py [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+    """Produce bar chart of numbers of households in each cluster.
+
+    Args:
+        data (pd.DataFrame): Household usage data.
+        clusters (list): Cluster designations.
+        filename_infix (str): Description of variant (e.g. "normalised_usage").
+            Appears in filename.
+    """
+    data["cluster"] = clusters
+    counts = data["cluster"].value_counts().reset_index(name="count")
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/utils/plotting_utils.py]: we don't really need the `data`, do we? The `clusters` variable seem to have all the information, or did I get this wrong?

File: asf_smart_meter_exploration/pipeline/data_aggregation.py [hunk 25]
```diff
@@ -0,0 +1,177 @@
...
+        seasonal_diff = (seasonal_aves.loc[season_dict[season_1]] - spring_autumn_ave).T
+
+    return seasonal_diff
+
+
+def merge_household_data(usage_data):
+    """Merge household contextual data (tariff and Acorn group) onto usage dataframe.
+
+    Args:
+        usage_data (pd.DataFrame): Smart meter data where index is household ID.
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/pipeline/data_aggregation.py:163]: I'm guessing we don't use the `meter_data` as default because we are doing the merge using the different aggregated dataframes.

File: asf_smart_meter_exploration/utils/clustering_utils.py [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+import matplotlib.pyplot as plt
+
+from asf_smart_meter_exploration import config, PROJECT_DIR
+
+inertia_plot_folder_path = PROJECT_DIR / config["inertia_plot_folder_path"]
+plot_suffix = config["plot_suffix"]
+random_state = config["random_state"]
+
+
+def inertia_plot(data, filename, n=10):
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/utils/clustering_utils.py]: I would probably split this into two functions: one that calculates the inertia across different values of k. Another to plot the inertia values across different values of k (this could live in the `plotting_utils`).

File: asf_smart_meter_exploration/utils/clustering_utils.py [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+            columns for each half hour.
+        n (int, optional): Max k to try. Defaults to 10.
+    """
+
+    inertias = []
+
+    for i in range(1, n + 1):
+        kmeans = KMeans(n_clusters=i, random_state=random_state, n_init=10)
+        kmeans.fit(data)
+        inertias.append(kmeans.inertia_)
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/utils/clustering_utils.py]: Important to bear in mind that `kmeans.inertia_` is the best output of 10 consecutive runs in terms of inertia, and not the average inertia across the 10 runs.

Also, why are we setting a random state at this point? It might be premature and bias the results.

File: asf_smart_meter_exploration/analysis/clustering.py [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+        # Create and save cluster plots using the parameters from the dict
+        # Use variant name as part of filename
+        plot_observations_and_clusters(
+            df,
+            clusters,
+            filename_infix=type,
+            normalised=type_dict["normalised"],
+            ylabel=type_dict["ylabel"],
+            ymin=type_dict["ymin"],
+            ymax=type_dict["ymax"],
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/analysis/clustering.py:53]: Do we really need to set this _a priori_? I think we can calculate ymin and ymax based on the data.

File: asf_smart_meter_exploration/analysis/clustering.py [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+from asf_smart_meter_exploration.utils.plotting_utils import *
+from asf_smart_meter_exploration.pipeline.data_aggregation import *
+
+# Dictionary of variants to cluster and plot.
+# "df" is the smart meter dataframe, "k" is the number of clusters,
+# and see docs for `plot_observations_and_clusters` (in `utils/plotting_utils.py`)
+# for other parameters.
+# Values of k here were chosen after analysing plots produced in
+# `analysis/inertia_plots.py`.
+options_dict = {
```
Comment (sofiapinto) [inline asf_smart_meter_exploration/analysis/clustering.py]: Not sure whether this dictionary should be in a config file or here (according to [this](https://docs.google.com/presentation/d/1F2vYPstP2sRAysvmnK0wHpCBTl-Vou_m9jraN7vX4E0/edit#slide=id.g175b4677f73_0_233))

I think maybe a config file? Otherwise the `inertia_plots.py` becomes dependent on this, and all modules in analysis should be independent of each other as per the slide deck above.

**Review by sofiapinto — APPROVED** [review]
Hi @ch-williamson ,

Thanks a lot for this! The code is very neat and I've learned a few pandas functions I've not used before :) 

Here's what I did besides reviewing the scripts:
- Ran `python3 asf_smart_meter_exploration/getters/process_raw_data.py` which stores a processed dataset, `electricity_data.csv`, under `outputs/data/`. When doing this I also get a `__MACOSX` folder with an empty `halfhourly_dataset` folder inside (under `inputs`). I left a few suggestions on how I think we could improve and make this script a little faster :)
- Ran `python3 asf_smart_meter_exploration/analysis/clustering.py` which produces 29 figures under `outputs/figures/clusters/`. I wonder if this is the best script to define `options_dict` (?) - left a comment about it. Also it would probably be better not to set a random state when doing the inertia plots (in `asf_smart_meter_exploration/utils/clustering_utils.py`).
- Ran `python3 asf_smart_meter_exploration/analysis/inertia_plots.py` which produces 8 figures under `/outputs/figures/inertia/`.

Not mandatory, but since this is an exploratory project, it would be nice to have a notebook where people could take a look at the raw data, processed data, aggregated datasets, etc. It would basically just call some of the functions you have already created.

Most of my comments are just suggestions, so feel free to ignore whatever does not make sense to you.

File: README.md (hunk 2 above)
Comment (ch-williamson) [reply README.md]: Good point! I'll just remove the second part because the script should cover the unzipping bit

File: asf_smart_meter_exploration/analysis/clustering.py (hunk 29 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/analysis/clustering.py]: Agree, keeping it in a config file is better!

File: asf_smart_meter_exploration/analysis/clustering.py (hunk 28 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/analysis/clustering.py:53]: Ideally it would be based on the data but since we're plotting all the individual lines there tend to be outliers that mess up the scales! It ended up being easier to specify ymin/ymax based on what needs to be shown

Comment (sofiapinto) [reply asf_smart_meter_exploration/analysis/clustering.py:53]: Makes sense @ch-williamson :)

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 18 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py]: Good idea, changed!

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 19 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py]: Mainly because this was the only combination I was interested in, so didn't want to spend the time making it more general! The reason being that seasonal variation is mainly due to heating/cooling, so we would expect spring and autumn to be pretty similar and it makes more sense to combine them than any other two seasons. I might not generalise this now but definitely something that could be done later

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 17 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py]: Thank you!

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 16 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py]: Thank you :)

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 12 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py]: I think the missing values are genuine errors rather than no electricity being used. But agree that we should look out for this in the future and more rigorously check proportions of missing values (I think this is relatively clean/complete data as far as smart meter data goes so we will probably have to do a lot more work on this when we get a different dataset)

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 15 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py:83]: I included this just in case there were households that used no electricity and we ended up with 0/0 = NaN. Not sure there are any here but maybe worth keeping so this can be adapted for other datasets

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 13 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py:32]: See other comment - this is in case of 0/0 = NaN

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 14 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py:76]: This was in case some households had no readings for a particular daytype/time combination - these wouldn't necessarily have been removed already

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 25 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py:163]: That's right :)

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 20 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py:1]: Great spot, this is dodgy. I think I either need to rename the `meter_data` variables inside the functions to ensure that they're not editing it, or explicitly `.copy()`

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 11 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py]: Added everywhere, thanks!

File: asf_smart_meter_exploration/utils/clustering_utils.py (hunk 26 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/utils/clustering_utils.py]: Good idea - done

File: asf_smart_meter_exploration/utils/clustering_utils.py (hunk 27 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/utils/clustering_utils.py]: Good point! Yes, I added this for reproducability but it defeats the point - I'll remove it

File: asf_smart_meter_exploration/utils/plotting_utils.py (hunk 21 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/utils/plotting_utils.py]: Good idea :) changed

File: asf_smart_meter_exploration/utils/plotting_utils.py (hunk 23 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/utils/plotting_utils.py:112]: I think it depends on the context - not completely sure why but when the `.clf()` line wasn't here my plots would occasionally overwrite each other. Will leave in since it's harmless but thanks for raising!

File: asf_smart_meter_exploration/utils/plotting_utils.py (hunk 24 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/utils/plotting_utils.py]: Good point!

File: requirements.txt (hunk 10 above)
Comment (ch-williamson) [reply requirements.txt:9]: I added it while trying to get altair_saver to work, but I think I ended up using a different method, so can probably remove this! Will remove and test

File: asf_smart_meter_exploration/pipeline/data_aggregation.py (hunk 20 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/pipeline/data_aggregation.py:1]: Fixed this

File: asf_smart_meter_exploration/utils/plotting_utils.py (hunk 22 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/utils/plotting_utils.py]: Weird! Possibly because with certain ymin/ymax values, 4 gives awkward numbers so it uses something close to 4 instead? I'm not sure this line is even really needed as the default bins are fine, so will remove

File: README.md (hunk 8 above)
Comment (ch-williamson) [reply README.md]: Good suggestion - have done this :)

File: requirements.txt (hunk 10 above)
Comment (ch-williamson) [reply requirements.txt:9]: Actually it seems to be needed for saving the plots so will leave it in!

File: asf_smart_meter_exploration/getters/process_raw_data.py (hunk 3 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/getters/process_raw_data.py]: Have made these changes - looks good, thanks :)

File: asf_smart_meter_exploration/getters/process_raw_data.py (hunk 6 above)
Comment (ch-williamson) [reply asf_smart_meter_exploration/getters/process_raw_data.py]: I think what's actually happening here is that there are a few records in the original data where the readings are the string "Null", and these are being removed (the original data structure is one row for each property and timestamp so the rest of the readings for that property are kept). So I think this is okay, but we can be a lot more careful about this when we have better data!

File: asf_smart_meter_exploration/utils/plotting_utils.py (hunk 23 above)
Comment (sofiapinto) [reply asf_smart_meter_exploration/utils/plotting_utils.py:112]: Good to know, thanks @ch-williamson :)

**Comment by sofiapinto** [comment]
Let me know if you need me to run the scripts again after the changes to ensure everything still runs smoothly @ch-williamson :)

**Comment by ch-williamson** [comment]
Thanks so much @sofiapinto, that would be great if you don't mind! I've actioned your suggestions and added a notebook for testing the main functions - it's ready for you to look at whenever you're free :)

**Comment by sofiapinto** [comment]
Hey @ch-williamson,

I've re-ran:
- `python3 asf_smart_meter_exploration/analysis/clustering.py`
- `python3 asf_smart_meter_exploration/analysis/inertia_plots.py`

Now we no longer need to run the `process_raw_data.py` script separately, so I didn't do that. Everything ran smoothly, as before.

Only got this warning, which also appeared before when running the clustering.py script:
```
Processing the data...
100%|█████████████████████████████████████████| 112/112 [29:54<00:00, 16.02s/it]
/Users/anasofiapinto/opt/anaconda3/envs/asf_smart_meter_exploration/lib/python3.8/site-packages/pandas/core/internals/blocks.py:2323: RuntimeWarning: invalid value encountered in cast
  values = values.astype(str)
```
It doesn't stop the code from running, but probably good to check what's happening.

Also took a look at the new notebook you created. Looks good and it's very useful - thanks for adding it!

Thanks so much for your work.

**Comment by ch-williamson** [comment]
Hmm, I can't reproduce this 🤔 From a bit of googling it looks like it's something to do with the NaNs in the dataframe, but it might depend on the versions of pandas/Python that are being used, or even the environment it's being run in. I'll leave it for now since everything works otherwise. Thanks so much for your review!
